File: 백준/Gold/2073. 수도배관공사/수도배관공사.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int[] dp = new int[N+1];
        dp[0] = Integer.MAX_VALUE;
        for (int i=0;i<M;i++) {
            st = new StringTokenizer(br.readLine());
            int L = Integer.parseInt(st.nextToken());
            int C = Integer.parseInt(st.nextToken());
            for (int j=N;j>=L;j--){
                dp[j] = Math.max(dp[j], Math.min(C, dp[j-L]));                
            }
        }
        System.out.println(dp[N]);
    }
}

File: 백준/Gold/2073. 수도배관공사/수도배관공사.py
D,P = map(int,input().split())
dp = [1e9]+[0]*(D)
arr = []
for _ in range(P):
    L,C = map(int,input().split())
    for i in range(D,L-1,-1):
        dp[i] = max(dp[i],min(C,dp[i-L]))

print(dp[-1])